Sources/Basics/HTTPClient/URLSessionHTTPClient.swift

```diff
@@ -337,6 +353,270 @@ private class DownloadTaskManager: NSObject, URLSessionDownloadDelegate {
     }
 }
 
+private class UploadTaskManager: NSObject, URLSessionTaskDelegate, URLSessionDataDelegate {
+    private var tasks = ThreadSafeKeyValueStore<Int, UploadTask>()
+    private let delegateQueue: OperationQueue
+    private var session: URLSession!
+
+    init(configuration: URLSessionConfiguration) {
+        self.delegateQueue = OperationQueue()
+        self.delegateQueue.name = "org.swift.swiftpm.urlsession-http-client-upload-delegate"
+        self.delegateQueue.maxConcurrentOperationCount = 1
+        super.init()
+        self.session = URLSession(configuration: configuration, delegate: self, delegateQueue: self.delegateQueue)
+    }
+
+    func makeTask(
+        urlRequest: URLRequest,
+        streamProvider: @escaping () throws -> LegacyHTTPClientRequest.UploadData?,
+        progress: LegacyHTTPClient.ProgressHandler?,
+        completion: @escaping LegacyHTTPClient.CompletionHandler
+    ) throws -> URLSessionUploadTask {
+        let task = self.session.uploadTask(withStreamedRequest: urlRequest)
+        self.tasks[task.taskIdentifier] = try UploadTask(
+            underlying: task,
+            streamProvider: streamProvider,
+            progressHandler: progress,
+            completionHandler: completion
+        )
+        return task
+    }
+
+    func urlSession(
+        _ session: URLSession,
+        task: URLSessionTask,
+        needNewBodyStream completionHandler: @escaping @Sendable (InputStream?) -> Void
+    ) {
+        guard let task = self.tasks[task.taskIdentifier] else {
+            return
+        }
+        completionHandler(task.inputStream)
+        // FIXME: serious hack here
+        RunLoop.current.run(until: Date(timeIntervalSinceNow: 10000.0))
```

Comment:
@weissi any ideas on why the stream is not going notified unless I run the .current runloop manually?

Reply:
that almost certainly means that you're blocking the runloop somewhere.